src/lua/timer.rs

```diff
@@ -0,0 +1,255 @@
+use rlua::{UserData, UserDataMethods};
+
+use super::{
+    backend::Backend,
+    blight::Blight,
+    constants::{BACKEND, TIMED_CALLBACK_TABLE, TIMED_CALLBACK_TABLE_CORE},
+};
+use crate::event::Event;
+use chrono::Duration;
+use std::error::Error;
+use std::sync::Arc;
+
+pub struct Timer {
+    next_id: u32,
+}
+
+impl Timer {
+    pub fn new() -> Self {
+        Self { next_id: 0 }
+    }
+
+    fn next_index(&mut self) -> u32 {
+        self.next_id += 1;
+        self.next_id
+    }
+}
+
+fn is_core_mode(ctx: rlua::Context) -> Result<bool, rlua::Error> {
+    let blight: Blight = ctx.globals().get("blight")?;
+    Ok(blight.core_mode)
+}
+
+fn user_mode_only(ctx: rlua::Context) -> Result<(), rlua::Error> {
+    if is_core_mode(ctx)? {
+        let boxed_error = Box::<dyn Error + Send + Sync>::from(NOT_SUPPORTED_IN_CORE_MODE);
+        return Err(rlua::Error::ExternalError(Arc::from(boxed_error)));
+    }
+    return Ok(());
+}
+
+const NOT_SUPPORTED_IN_CORE_MODE: &str = "this method is not supported in core mode";
+
+impl UserData for Timer {
+    fn add_methods<'lua, M: UserDataMethods<'lua, Self>>(methods: &mut M) {
+        methods.add_method_mut(
```

Comment:
All occurrences of "add_method" can be replaced by "add_function". This is the only method that uses 'this' and after you've removed 'next_index()' you won't need it here either.